MSP-hands-on-code/papi-4.1.3/man/man3/PAPI_set_domain.3
.\" $Id: PAPI_set_domain.3,v 1.14 2009/09/10 20:19:38 terpstra Exp $
.TH PAPI_set_domain 3 "September, 2004" "PAPI Programmer's Reference" "PAPI"

.SH NAME
 PAPI_set_domain \- set the default counting domain for new event sets bound to the cpu component
 PAPI_set_cmp_domain \- set the default counting domain for new event sets bound to the specified component

.SH SYNOPSIS
.B C Interface
.nf
.B #include <papi.h>
.BI "int PAPI_set_domain(int " domain ");"
.BI "int PAPI_set_cmp_domain(int " domain ", int " cidx ");"
.fi
.B Fortran Interface
.nf
.B #include "fpapi.h"
.BI PAPIF_set_domain(C_INT\  domain,\  C_INT\  check )
.BI PAPIF_set_event_domain(C_INT\  EventSet,\  C_INT\  domain,\  C_INT\  check )
.fi

.SH DESCRIPTION
.B "PAPI_set_domain" 
sets the default counting domain for all new event sets created by
.BR "PAPI_create_eventset" (3) 
in all threads. This call implicitly sets the domain for the cpu component
(component 0) and is included to preserve backward compatibility.
.LP
.B "PAPI_set_cmp_domain" 
sets the default counting domain for all new event sets
in all threads, and requires an explicit component argument. 
Event sets that are already in existence are not affected. 
To change the domain of an existing event set, please see the 
.BR "PAPI_set_opt" (3)
man page.The reader should note that the domain of an event set
affects only the mode in which the counter continues to run. Counts are still
aggregated for the current process, and not for any other processes in
the system. Thus when requesting
.BR PAPI_DOM_KERNEL , 
the user is asking for events that occur on behalf of the process,
inside the kernel.

.SH ARGUMENTS
.I cidx
-- An integer identifier for a component. By convention, component 0 is always the cpu component.

.I "domain"
-- one of the following constants as defined in the papi.h header file:

.TS
allbox tab($);
lB l.
PAPI_DOM_USER$User context counted
PAPI_DOM_KERNEL$Kernel/OS context counted
PAPI_DOM_OTHER$Exception/transient mode counted
PAPI_DOM_SUPERVISOR$Supervisor/hypervisor context counted
PAPI_DOM_ALL$All above contexts counted
PAPI_DOM_MIN$The smallest available context
PAPI_DOM_MAX$The largest available context
.TE

.SH RETURN VALUES
On success, this function returns
.B "PAPI_OK."
 On error, a non-zero error code is returned.

.SH ERRORS
.TP
.B "PAPI_EINVAL"
One or more of the arguments is invalid.
.TP
.B "PAPI_ENOEVST"
The event set specified does not exist.
.TP
.B "PAPI_ENOCMP"
The argument
.I cidx
is not a valid component.
.TP
.B "PAPI_EISRUN"
The event set is currently counting events.

.SH EXAMPLES
.nf
.if t .ft CW
int retval;

/* Initialize the library */

retval = PAPI_library_init(PAPI_VER_CURRENT);

if (retval > 0 && retval != PAPI_VER_CURRENT) {
  fprintf(stderr,"PAPI library version mismatch!\n");
  exit(1); }

if (retval < 0) 
  handle_error(retval);

/* Set the default domain for the cpu component */

if ((retval = PAPI_set_cmp_domain(PAPI_DOM_KERNEL,0)) != PAPI_OK)
  handle_error(retval);

if ((retval = PAPI_create_eventset(&EventSet)) != PAPI_OK)
  handle_error(retval);
.if t .ft P
.fi

.SH BUGS
This function has no known bugs.

.SH SEE ALSO
.BR PAPI_set_granularity "(3), " PAPI_set_opt "(3), " PAPI_get_opt "(3) "


MSP-hands-on-code/papi-4.1.3/src/examples/PAPI_set_domain.c
/*****************************************************************************
 * This example shows how to use PAPI_set_domain                             * 
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#include "papi.h" /* This needs to be included every time you use PAPI */

#define ERROR_RETURN(retval) { fprintf(stderr, "Error %d %s:line %d: \n", retval,__FILE__,__LINE__);  exit(retval); }

int poorly_tuned_function()
{
   float tmp;
   int i;

   for(i=1; i<2000; i++)
   {
      tmp=(tmp+100)/i;
   }
   return 0;
}

int main()
{

   int num, retval, EventSet = PAPI_NULL;
   long long values[2];
   PAPI_option_t options;    
   int fd;
   

   /****************************************************************************
   *  This part initializes the library and compares the version number of the *
   * header file, to the version of the library, if these don't match then it  *
   * is likely that PAPI won't work correctly.If there is an error, retval     *
   * keeps track of the version number.                                        *
   ****************************************************************************/

   if((retval = PAPI_library_init(PAPI_VER_CURRENT)) != PAPI_VER_CURRENT )
   {
      printf("Library initialization error! \n");
      exit(1);
   }

   /* Set the domain of this EventSet to counter user mode. The domain
      will be valid for all the eventset created after this function call 
      unless you call PAPI_set_domain again */ 
   if ((retval=PAPI_set_domain(PAPI_DOM_USER)) != PAPI_OK)
      ERROR_RETURN(retval);

   if ((retval=PAPI_create_eventset(&EventSet)) != PAPI_OK)
      ERROR_RETURN(retval);

   /* Add Total Instructions Executed event to the EventSet */
   if ( (retval = PAPI_add_event(EventSet, PAPI_TOT_INS)) != PAPI_OK)
      ERROR_RETURN(retval);

   /* Add Total Cycles Executed event to the EventSet */
   if ( (retval = PAPI_add_event(EventSet, PAPI_TOT_CYC)) != PAPI_OK)
      ERROR_RETURN(retval);

   /* Start counting */
   if((retval=PAPI_start(EventSet)) != PAPI_OK)
      ERROR_RETURN(retval);

   poorly_tuned_function();
   /* add some system calls */
   fd = open("/dev/zero", O_RDONLY);
   if (fd == -1)
   {
         perror("open(/dev/zero)");
         exit(1);
   }
   close(fd);


   /* Stop counting */
   if((retval=PAPI_stop(EventSet, values)) != PAPI_OK)
      ERROR_RETURN(retval);
        
   printf(" Total instructions: %lld   Total Cycles: %lld \n", values[0],
            values[1]);

   /* Set the domain of this EventSet to counter user and kernel modes */ 
   if ((retval=PAPI_set_domain(PAPI_DOM_ALL)) != PAPI_OK)
      ERROR_RETURN(retval);

   EventSet = PAPI_NULL;
   if ((retval=PAPI_create_eventset(&EventSet)) != PAPI_OK)
      ERROR_RETURN(retval);

   /* Add Total Instructions Executed to our EventSet */
   if ( (retval = PAPI_add_event(EventSet, PAPI_TOT_INS)) != PAPI_OK)
      ERROR_RETURN(retval);

   /* Add Total Instructions Executed to our EventSet */
   if ( (retval = PAPI_add_event(EventSet, PAPI_TOT_CYC)) != PAPI_OK)
      ERROR_RETURN(retval);
   /* Start counting */
   if((retval=PAPI_start(EventSet)) != PAPI_OK)
      ERROR_RETURN(retval);

   poorly_tuned_function();
   /* add some system calls */
   fd = open("/dev/zero", O_RDONLY);
   if (fd == -1)
   {
         perror("open(/dev/zero)");
         exit(1);
   }
   close(fd);

   /* Stop counting */
   if((retval=PAPI_stop(EventSet, values)) != PAPI_OK)
      ERROR_RETURN(retval);
        
   printf(" Total instructions: %lld   Total Cycles: %lld \n", values[0],
            values[1]);

   /* clean up */
   PAPI_shutdown();

   exit(0);
}
